### ISSUE TITLE
Desired capability `platform: windows` doesn't work

### ISSUE THREAD

**Issue by kottenator** [mozilla/geckodriver#1030]
## System
* Version: 0.19.0
* Platform: Windows Server 2012 R2
* Firefox: 56.0.1
* Selenium: 3.6.0

## Testcase

Launch `geckodriver` on Windows NT machine:

```sh
geckodriver.exe --log trace -b C:\Firefox\56.0.1\firefox.exe
```

And try to create a new session:

```sh
curl -i localhost:4444/session \
    -X POST -H 'Content-Type: application/json' \
    -d '{"desiredCapabilities": {"browserName": "firefox", "platformName": "windows"}}'
```

First, Geckodriver **will** open Firefox. But right after that - Firefox will crash and the following error will appear:

```
TypeError: Given platformName [object String] "windows", 
but current platform is [object String] "windows_nt"
```

FYI: I've tried to specify `"platformName": "windows_nt"` and `"platformName": "ANY"` - Geckodriver doesn't accept them.

## Stacktrace

```
stack backtrace:
    0: 0x4821c4 - <no info>
    1: 0x482933 - <no info>
    2: 0x445159 - <no info>
    3: 0x44eff6 - <no info>
    4: 0x4322e1 - <no info>
    5: 0x40ba4e - <no info>
    6: 0x4119b9 - <no info>
    7: 0x6d2c89 - <no info>
    8: 0x426006 - <no info>
    9: 0x6cd490 - <no info>
   10: 0x7ffee06a13d2 - BaseThreadInitThunk
```

**Comment by andreastt** [comment]
`desiredCapabilities`/`requiredCapabilities` is the deprecated Selenium-styled way of requesting capabilities. Have you tried with the WebDriver-styled capabilities as described in the README?

**Comment by kottenator** [comment]
@andreastt - I've tried the following:

```sh
curl -i localhost:4444/session -X POST -H 'Content-Type: application/json' -d '{"capabilities": {"browserName": "firefox", "platformName": "windows"}}'
```

And it worked. 

However, [Selenium 3.6.0 Python binding](https://github.com/SeleniumHQ/selenium/blob/selenium-3.6.0/py/selenium/webdriver/remote/webdriver.py#L240) generates the following request:

```
curl -i localhost:4444/session -X POST -H 'Content-Type: application/json' -d '{"capabilities": {"firstMatch": [{}], "alwaysMatch": {"browserName": "firefox", "platformName": "windows"}}, "desiredCapabilities": {"browserName": "firefox", "platform": "windows", "version": ""}}'
```

And it fails with the same error as described above. 

TL;DR: `{"capabilities": {"alwaysMatch": {...}}}` seems to be the problem. And it's [a part of W3C WebDriver spec](https://www.w3.org/TR/webdriver/#processing-capabilities) so probably you need to support it along with `{"capabilities": {...}}`.

**Comment by kottenator** [comment]
Here's a simpler example how you can reproduce the problem:

```sh
curl -i localhost:5555/session -X POST -H 'Content-Type: application/json' -d '{"capabilities": {"alwaysMatch": {"browserName": "firefox", "platformName": "windows"}}}'
```

**Comment by andreastt** [comment]
The HTTP request you use in your last comment looks correct. The ones mentioned earlier are either deprecated or wrong.

Because the platform is `windows_nt` you need to be more specific with your `platformName` choice for this to work with the current geckodriver.

AIUI it is correct for geckodriver/Marionette to _return_ `windows_nt`, which is a more specific platform name than `windows`, but the specification urges us to recognise `windows` on negotiation when [matching capabilities](https://w3c.github.io/webdriver/webdriver-spec.html#dfn-matching-capabilities) in step 3.

@jgraham: Am I reading this right?

**Comment by kottenator** [comment]
@andreastt - I've tried as you said:

> Because the platform is `windows_nt` you need to be more specific with your platformName choice for this to work with the current geckodriver.

And it gives me another error:

```
{"value":{"error":"session not created","message":"Unable to find a matching set of capabilities","stacktrace":"stack backtrace:\n   0:           0x4821c4 - <no info>\n   1:           0x482933 - <no info>\n   2:           0x445281 - <no info>\n   3:           0x4462ca - <no info>\n   4:           0x43026a - <no info>\n   5:           0x40ba4e - <no info>\n   6:           0x4119b9 - <no info>\n   7:           0x6d2c89 - <no info>\n   8:           0x426006 - <no info>\n   9:           0x6cd490 - <no info>\n  10:     0x7ffee06a13d2 - BaseThreadInitThunk"}}
```

**Comment by andreastt** [comment]
And the input for that?

**Comment by kottenator** [comment]
Here it is:
```sh
curl -i localhost:4444/session -X POST -H 'Content-Type: application/json' \
    -d '{"capabilities": {"alwaysMatch": {"browserName": "firefox", "platformName": "windows_nt"}}}'
```

**Comment by rgoerner** [comment]
I can confirm this. Geckodriver, executed over a Webdriver-Session and platformName ="windows" fails, because it requests "windows_nt" which is wrong. :-<

**Comment by andreastt** [comment]
@rgoerner It’s technically not against the specification, but we should probably recognise `windows` if you’re on a Windows system. It is obviously fine for it to return with `windows_nt` because that is more specific.

I have filed https://bugzilla.mozilla.org/show_bug.cgi?id=1414320 to follow up on this.

**Comment by andreastt** [comment]
@kottenator I can’t actually reproduce that problem on my Linux system, but that may be because the negotiated `platformName` there actually is `linux`. Would you mind running that again with a trace-level log so we can see what Marionette returns?

**Comment by kottenator** [comment]
@andreastt - sure, I'll gladly help you. Although, I'm not sure which case are you asking to redo, so I'll do two.

Request 1: `platformName: windows`

```sh
curl -i localhost:4444/session -X POST -H 'Content-Type: application/json' -d '{"capabilities": {"alwaysMatch": {"browserName": "firefox", "platformName": "windows"}, "moz:firefoxOptions": {"log": {"level": "TRACE"}}}}'
```

Output 1: `windows` != `windows_nt`

```
1509739086816   geckodriver     INFO    geckodriver 0.19.0
1509739086818   webdriver::httpapi      DEBUG   Creating routes
1509739086837   geckodriver     INFO    Listening on 127.0.0.1:4444
1509739095809   webdriver::server       DEBUG   -> POST /session {"capabilities": {"alwaysMatch": {"browserName": "firefox", "platformName": "windows"}, "moz:firefoxOptions": {"log": {"level": "TRACE"}}}
1509739095895   mozrunner::runner       INFO    Running command: "..\\..\\firefox\\56.0.1\\firefox.exe" "-marionette" "-profile" "C:\\Users\\bryzr\\AppData\\Local\\Temp\\51\\rust_mozprofile.NHI3SL8fB1gC"
1509739098014   geckodriver::marionette TRACE     connection attempt 0/600
1509739100123   geckodriver::marionette DEBUG   Connected to Marionette on localhost:65086
1509739100137   geckodriver::marionette TRACE   <- {"applicationType":"gecko","marionetteProtocol":3}
1509739100138   geckodriver::marionette TRACE   -> 159:[0,1,"newSession",{"browserName":"firefox","capabilities":{"desiredCapabilities":{"browserName":"firefox","platformName":"windows"}},"platformName":"windows"}]
1509739100152   geckodriver::marionette TRACE   <- [1,1,{"error":"session not created","message":"TypeError: Given platformName [object String] \"windows\", but current platform is [object String] \"windows_nt\"","stacktrace":"match_@chrome://marionette/content/session.js:384:19\nfromJSON@chrome://marionette/content/session.js:326:12\nGeckoDriver.prototype.newSession@chrome://marionette/content/driver.js:671:25\nTaskImpl_run@resource://gre/modules/Task.jsm:331:42\nTaskImpl@resource://gre/modules/Task.jsm:280:3\nasyncFunction@resource://gre/modules/Task.jsm:252:14\nTask_spawn@resource://gre/modules/Task.jsm:166:12\nTaskImpl_handleResultValue@resource://gre/modules/Task.jsm:401:16\nTaskImpl_run@resource://gre/modules/Task.jsm:339:15\nTaskImpl@resource://gre/modules/Task.jsm:280:3\nasyncFunction@resource://gre/modules/Task.jsm:252:14\nTask_spawn@resource://gre/modules/Task.jsm:166:12\nexecute@chrome://marionette/content/server.js:529:15\nonPacket@chrome://marionette/content/server.js:500:7\n_onJSONObjectReady/<chrome://marionette/content/transport.js:501:9\n"},null]
1509739100163   webdriver::server       DEBUG   Deleting session
1509739100163   geckodriver::marionette DEBUG   Stopping browser process
1509739100295   webdriver::server       DEBUG   <- 500 Internal Server Error {"value":{"error":"session not created","message":"TypeError: Given platformName [object String] \"windows\", but current platform is [object String] \"windows_nt\"","stacktrace":"stack backtrace:\n   0:           0x4821c4 - <no info>\n   1:         0x482933 - <no info>\n   2:           0x445159 - <no info>\n   3:     0x44eff6 - <no info>\n   4:           0x4322e1 - <no info>\n   5: 0x40ba4e - <no info>\n   6:           0x4119b9 - <no info>\n   7:           0x6d2c89 - <no info>\n   8:           0x426006 - <no info>\n   9:           0x6cd490 - <no info>\n  10:     0x7ffee06a13d2 - BaseThreadInitThunk"}}
````

Request 2: `platformName: windows_nt`

```sh
curl -i localhost:4444/session -X POST -H 'Content-Type: application/json' -d '{"capabilities": {"alwaysMatch": {"browserName": "firefox", "platformName": "windows_nt"}, "moz:firefoxOptions": {"log": {"level": "TRACE"}}}}'
```

Output 2: `windows_nt` not found

```
1509739821476   geckodriver     INFO    geckodriver 0.19.0
1509739821476   webdriver::httpapi      DEBUG   Creating routes
1509739821492   geckodriver     INFO    Listening on 127.0.0.1:4444
1509739843711   webdriver::server       DEBUG   -> POST /session {"capabilities": {"alwaysMatch": {"browserName": "firefox", "platformName": "windows_nt"}, "moz:firefoxOptions": {"log": {"level": "TRACE"}}}}
1509739843757   webdriver::server       DEBUG   <- 500 Internal Server Error {"value":{"error":"session not created","message":"Unable to find a matching set of capabilities","stacktrace":"stack backtrace:\n   0:           0x4821c4 - <no info>\n   1:           0x482933 - <no info>\n   2:           0x445281 - <no info>\n   3:           0x4462ca - <no info>\n   4:           0x43026a - <no info>\n5:           0x40ba4e - <no info>\n   6:           0x4119b9 - <no info>\n   7:         0x6d2c89 - <no info>\n   8:           0x426006 - <no info>\n   9:     0x6cd490 - <no info>\n  10:     0x7ffee06a13d2 - BaseThreadInitThunk"}}
```

**Comment by andreastt** [comment]
That’s really odd. Can you try once _without_ specifying `platformName`, but with a trace log?

**Comment by kottenator** [comment]
Request 3: without `platformName`

```sh
curl -i localhost:4444/session -X POST -H 'Content-Type: application/json' -d '{"capabilities": {"alwaysMatch": {"browserName": "firefox"}, "moz:firefoxOptions": {"log": {"level": "TRACE"}}}}'
```

Output 3: works fine

```
1509740169120   geckodriver     INFO    geckodriver 0.19.0
1509740169120   webdriver::httpapi      DEBUG   Creating routes
1509740169136   geckodriver     INFO    Listening on 127.0.0.1:4444
1509740180583   webdriver::server       DEBUG   -> POST /session {"capabilities": {"alwaysMatch": {"browserName": "firefox"}, "moz:firefoxOptions": {"log": {"level": "TRACE"}}}}
1509740180661   mozrunner::runner       INFO    Running command: "..\\..\\firefox\\56.0.1\\firefox.exe" "-marionette" "-profile" "C:\\Users\\bryzr\\AppData\\Local\\Temp\\51\\rust_mozprofile.0Kp40JbRd7ls"
1509740182824   geckodriver::marionette TRACE     connection attempt 0/600
1509740184978   geckodriver::marionette TRACE     connection attempt 1/600
1509740186117   geckodriver::marionette DEBUG   Connected to Marionette on localhost:57198
1509740186117   geckodriver::marionette TRACE   <- {"applicationType":"gecko","marionetteProtocol":3}
1509740186117   geckodriver::marionette TRACE   -> 109:[0,1,"newSession",{"browserName":"firefox","capabilities":{"desiredCapabilities":{"browserName":"firefox"}}}]
1509740186336   geckodriver::marionette TRACE   <- [1,1,null,{"sessionId":"328679f6-31a1-406c-a98d-dd49f033fa42","capabilities":{"browserName":"firefox","browserVersion":"56.0.1","platformName":"windowsnt","platformVersion":"6.3","pageLoadStrategy":"normal","acceptInsecureCerts":false,"timeouts":{"implicit":0,"pageLoad":300000,"script":30000},"rotatable":false,"specificationLevel":0,"moz:processID":319992,"moz:profile":"C:\\Users\\bryzr\\AppData\\Local\\Temp\\51\\rust_mozprofile.0Kp40JbRd7ls","moz:accessibilityChecks":false,"moz:headless":false}}]
1509740186336   webdriver::server       DEBUG   <- 200 OK {"value":{"sessionId":"328679f6-31a1-406c-a98d-dd49f033fa42","capabilities":{"acceptInsecureCerts":false,"browserName":"firefox","browserVersion":"56.0.1","moz:accessibilityChecks":false,"moz:headless":false,"moz:processID":319992,"moz:profile":"C:\\Users\\bryzr\\AppData\\Local\\Temp\\51\\rust_mozprofile.0Kp40JbRd7ls","pageLoadStrategy":"normal","platformName":"windows_nt","platformVersion":"6.3","rotatable":false,"specificationLevel":0,"timeouts":{"implicit":0,"pageLoad":300000,"script":30000}}}}
```

**Comment by andreastt** [comment]
Right, so the negotiated platform name on your system is `windowsnt` (without underscore). This should work for input until we fix https://bugzilla.mozilla.org/show_bug.cgi?id=1414320.

**Comment by kottenator** [comment]
Unfortunately, it doesn't work.

Request 4: `platformName: windowsnt`

```sh
curl -i localhost:4444/session -X POST -H 'Content-Type: application/json' -d '{"capabilities": {"alwaysMatch": {"browserName": "firefox", "platformName": "windowsnt"}, "moz:firefoxOptions": {"log": {"level": "TRACE"}}}}'
```

Output 4: `windowsnt` not found

```
1509741589161   geckodriver     INFO    geckodriver 0.19.0
1509741589164   webdriver::httpapi      DEBUG   Creating routes
1509741589184   geckodriver     INFO    Listening on 127.0.0.1:4444
1509741602680   webdriver::server       DEBUG   -> POST /session {"capabilities": {"alwaysMatch": {"browserName": "firefox", "platformName": "windowsnt"}, "moz:firefoxOptions": {"log": {"level": "TRACE"}}}}
1509741602738   webdriver::server       DEBUG   <- 500 Internal Server Error {"value":{"error":"session not created","message":"Unable to find a matching set of capabilities","stacktrace":"stack backtrace:\n   0:           0x4821c4 - <no info>\n   1:           0x482933 - <no info>\n   2:           0x445281 - <no info>\n   3:           0x4462ca - <no info>\n   4:           0x43026a - <no info>\n5:           0x40ba4e - <no info>\n   6:           0x4119b9 - <no info>\n   7:         0x6d2c89 - <no info>\n   8:           0x426006 - <no info>\n   9:     0x6cd490 - <no info>\n  10:     0x7ffee06a13d2 - BaseThreadInitThunk"}}
```

**Comment by morgwai** [comment]
I was also hit by that issue. In case of my selenium grid which consists of a win10 node and a linux node this bug makes it impossible to request a firefox session on the windows node via java api.

**Comment by drptbl** [comment]
Same issue here.

**Comment by andreastt** [comment]
Yes, pass in `windows_nt`.

**Comment by drptbl** [comment]
@andreastt It was already stated above that this doesn't work or help in this case. And I can confirm that. Neither `windows_nt`, neither `windowsnt`.

**Comment by shs96c** [comment]
Watching this issue. For maximum interior, @andreastt is correct that the spec urges implementors to recognise "`windows`" as a valid platform name for any of the Windows family.

**Comment by AutomatedTester** [comment]
I've raised https://bugzilla.mozilla.org/show_bug.cgi?id=1421233

**Comment by andreastt** [comment]
@AutomatedTester I already raised https://bugzilla.mozilla.org/show_bug.cgi?id=1414320.

**Comment by andreastt** [comment]
As @jgraham noted in https://bugzilla.mozilla.org/show_bug.cgi?id=1414320#c2 this also requires a little bit of spec work.

**Comment by hon2a** [comment]
@andreastt We have `platform` explicitly set in node config (e.g. to `WIN8_1`) and we're experiencing this issue anyway, so this looks like a general issue with matching capabilities rather than a question of (not) recognizing `windows` for a Windows Server by default.

**Comment by whimboo** [comment]
With the patch on [bug 1470646](https://bugzilla.mozilla.org/show_bug.cgi?id=1470646) landed in Firefox 63 (Nightly) we now correctly report the platform name as specified in the WebDriver spec.

@andreastt is there anything else which is missing here?

**Comment by andreastt** [comment]
@whimboo Nope, I think that fixes it.

**Comment by whimboo** [comment]
> Well actually, https://bugzilla.mozilla.org/show_bug.cgi?id=1470646 has no impact on geckodriver: it only changes what is returned by Marionette.

Huh? That change fixed the wdspec tests which test for the platform name. So how could that not have an impact on geckodriver, which simply pass-through the platform name?

**Comment by andreastt** [comment]
Hm, you’re correct. We take what Marionette gives for the _return value_, but for input I can find no sign of validation.

But since we’re returning the correct values now, it was the right call to close this issue.
